# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [1]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.1.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.245882  128.789602  125.589255  127.168169  1.120920e+08
std      30.329168   30.605763   30.048110   30.357183  5.602586e+07
min      54.213600   55.211516   51.372081   55.110034  3.519590e+07
25%     112.367579  113.858474  110.914304  112.547194  7.636470e+07
50%     132.607239  133.736887  130.653095  132.429295  9.493580e+07
75%     147.771828  148.935375  145.833106  147.141023  1.296327e+08
max     178.103653  179.013714  175.275693  178.710370  4.265100e+08


In [2]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



[notice] A new release of pip is available: 25.1.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip
[*********************100%***********************]  1 of 1 completed

Defaulting to user installation because normal site-packages is not writeable
Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.245882  128.789602  125.589255  127.168169  1.120920e+08
std      30.329168   30.605763   30.048110   30.357183  5.602586e+07
min      54.213600   55.211516   51.372081   55.110034  3.519590e+07
25%     112.367579  113.858474  110.914304  112.547194  7.636470e+07
50%     132.607239  133.736887  130.653095  132.429295  9.493580e+07
75%     147.771828  148.935375  145.833106  147.141023  1.296327e+08
max     178.103653  179.013714  175.275693  178.710370  4.265100e+08


In [3]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.245882  128.789602  125.589255  127.168169  1.120920e+08
std      30.329168   30.605763   30.048110   30.357183  5.602586e+07
min      54.213600   55.211516   51.372081   55.110034  3.519590e+07
25%     112.367579  113.858474  110.914304  112.547194  7.636470e+07
50%     132.607239  133.736887  130.653095  132.429295  9.493580e+07
75%     147.771828  148.935375  145.833106  147.141023  1.296327e+08
max     178.103653  179.013714  175.275693  178.710370  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



In [4]:
# Prompt: Give me the code that runs Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.
import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

[*********************100%***********************]  1 of 1 completed


In [5]:
# Summarize specific key statistics as requested
# Ensure 'Close' column exists and is numeric
if 'Close' in data.columns:
    highest_close = data['Close'].max()
    lowest_close = data['Close'].min()
    average_close = data['Close'].mean()
    total_trading_days = len(data)

    # Extract scalar values from Series for printing
    print(f"- **Highest Closing Price:** ${highest_close.iloc[0]:.2f}")
    print(f"- **Lowest Closing Price:** ${lowest_close.iloc[0]:.2f}")
    print(f"- **Average Closing Price:** ${average_close.iloc[0]:.2f}")
    print(f"- **Total Trading Days:** {total_trading_days}")
else:
    print("Close price data not found in the downloaded data.")

- **Highest Closing Price:** $178.10
- **Lowest Closing Price:** $54.21
- **Average Closing Price:** $127.25
- **Total Trading Days:** 756


In [6]:
#Prompt: Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.
!pip install ta

import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import ta
import pandas as pd

# Flatten MultiIndex columns if they exist. This is a common issue with yfinance data.
# Assuming columns are like ('Close', 'AAPL'), ('High', 'AAPL'), etc.
# We want to keep only the first level ('Close', 'High').
if isinstance(data.columns, pd.MultiIndex):
    data.columns = data.columns.droplevel(1)

# Ensure all necessary features are calculated
# Calculate Moving Averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_20'] = data['Close'].rolling(window=20).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Calculate Relative Strength Index (RSI)
data['RSI'] = ta.momentum.rsi(data['Close'], window=14)

# Calculate Moving Average Convergence Divergence (MACD)
data['MACD'] = ta.trend.macd(data['Close'])
data['MACD_Signal'] = ta.trend.macd_signal(data['Close'])
data['MACD_Diff'] = ta.trend.macd_diff(data['Close'])

# Create the target variable (tomorrow's closing price)
data['Tomorrow_Close'] = data['Close'].shift(-1)

# Define features (X) and target (y)
# First four kinds of features: Historical Prices, Volume, Moving Averages, Technical Indicators
features = [
    'Open', 'High', 'Low', 'Close', 'Volume',
    'MA_5', 'MA_10', 'MA_20', 'MA_50',
    'RSI', 'MACD', 'MACD_Signal', 'MACD_Diff'
]

X = data[features]
y = data['Tomorrow_Close']

# Drop rows with NaN values created by shifting and rolling window calculations
X = X.dropna()
y = y.dropna()

# Ensure X and y have the same number of samples after dropping NaNs
# This handles cases where dropping NaNs in X or y might result in different lengths
common_index = X.index.intersection(y.index)
X = X.loc[common_index]
y = y.loc[common_index]

# Split the data into training and testing sets (70/30 ratio)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Report performance metrics
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"### Regression Model Performance Metrics:\n")
print(f"- **Mean Squared Error (MSE):** {mse:.2f}")
print(f"- **R-squared (R²) Score:** {r2:.3f}")

print("\nThese metrics indicate how well the model explains the variance in tomorrow's stock price. A lower MSE is better, and an R² closer to 1 indicates a better fit. A negative R² suggests the model performs worse than simply predicting the mean of the target variable.")

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.1.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


ModuleNotFoundError: No module named 'sklearn'

In [ ]:
# Prompt: Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import pandas as pd

# Create the binary target variable
# 1 if tomorrow's close price is higher than today's close price, 0 otherwise
data['Tomorrow_Close_Direction'] = (data['Close'].shift(-1) > data['Close']).astype(int)

# Define features (X) and target (y)
# Using the same features as the regression model
features = [
    'Open', 'High', 'Low', 'Close', 'Volume',
    'MA_5', 'MA_10', 'MA_20', 'MA_50',
    'RSI', 'MACD', 'MACD_Signal', 'MACD_Diff'
]

X = data[features]
y = data['Tomorrow_Close_Direction']

# Drop rows with NaN values created by shifting and rolling window calculations
X = X.dropna()
y = y.dropna()

# Ensure X and y have the same number of samples after dropping NaNs
common_index = X.index.intersection(y.index)
X = X.loc[common_index]
y = y.loc[common_index]

# Split the data into training and testing sets (70/30 ratio)
# Using random_state for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Initialize and train the Random Forest Classifier model
model_rf = RandomForestClassifier(n_estimators=100, random_state=42)
model_rf.fit(X_train, y_train)

# Make predictions on the test set
y_pred_rf = model_rf.predict(X_test)

# Report performance metrics
accuracy_rf = accuracy_score(y_test, y_pred_rf)
conf_matrix_rf = confusion_matrix(y_test, y_pred_rf)
class_report_rf = classification_report(y_test, y_pred_rf)

print(f"### Random Forest Classifier Performance Metrics:\n")
print(f"- **Accuracy:** {accuracy_rf:.2f}")
print(f"- **Confusion Matrix:**\n{conf_matrix_rf}")
print(f"- **Classification Report:**\n{class_report_rf}")

# Report Feature Importance
feature_importance = pd.Series(model_rf.feature_importances_, index=features).sort_values(ascending=False)
print(f"\n### Feature Importance:\n{feature_importance}")

### Random Forest Classifier Performance Metrics:

- **Accuracy:** 0.53
- **Confusion Matrix:**
[[50 54]
 [46 63]]
- **Classification Report:**
              precision    recall  f1-score   support

           0       0.52      0.48      0.50       104
           1       0.54      0.58      0.56       109

    accuracy                           0.53       213
   macro avg       0.53      0.53      0.53       213
weighted avg       0.53      0.53      0.53       213


### Feature Importance:
Volume         0.089922
RSI            0.082938
Open           0.082241
MACD_Diff      0.081991
MA_5           0.077607
High           0.076776
Close          0.076123
MACD_Signal    0.073141
Low            0.072406
MA_50          0.072222
MACD           0.071933
MA_10          0.071619
MA_20          0.071079
dtype: float64


In [ ]:
#Prompt: Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.
import yfinance as yf
import pandas as pd
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import ta

# --- Step 1: Download new historical data for the prediction period ---
# Define the new prediction period
start_date_new = "2023-01-01"
end_date_new = "2024-01-01"

print(f"Downloading AAPL data for {start_date_new} to {end_date_new}...")
data_new = yf.download("AAPL", start=start_date_new, end=end_date_new)

# Ensure MultiIndex columns are handled if they appear in the new data
if isinstance(data_new.columns, pd.MultiIndex):
    data_new.columns = data_new.columns.droplevel(1)

# --- Step 2: Feature Engineering (same as training data) ---
# Calculate Moving Averages
data_new['MA_5'] = data_new['Close'].rolling(window=5).mean()
data_new['MA_10'] = data_new['Close'].rolling(window=10).mean()
data_new['MA_20'] = data_new['Close'].rolling(window=20).mean()
data_new['MA_50'] = data_new['Close'].rolling(window=50).mean()

# Calculate Relative Strength Index (RSI)
data_new['RSI'] = ta.momentum.rsi(data_new['Close'], window=14)

# Calculate Moving Average Convergence Divergence (MACD)
data_new['MACD'] = ta.trend.macd(data_new['Close'])
data_new['MACD_Signal'] = ta.trend.macd_signal(data_new['Close'])
data_new['MACD_Diff'] = ta.trend.macd_diff(data_new['Close'])

# Create the binary target variable for the new period
# 1 if tomorrow's close price is higher than today's close price, 0 otherwise
data_new['Tomorrow_Close_Direction'] = (data_new['Close'].shift(-1) > data_new['Close']).astype(int)

# Define features (X_new) and actual target (y_new) for the prediction period
features = [
    'Open', 'High', 'Low', 'Close', 'Volume',
    'MA_5', 'MA_10', 'MA_20', 'MA_50',
    'RSI', 'MACD', 'MACD_Signal', 'MACD_Diff'
]

X_new = data_new[features]
y_new = data_new['Tomorrow_Close_Direction']

# Drop rows with NaN values created by shifting and rolling window calculations
X_new = X_new.dropna()
y_new = y_new.dropna()

# Ensure X_new and y_new have the same number of samples after dropping NaNs
common_index_new = X_new.index.intersection(y_new.index)
X_new = X_new.loc[common_index_new]
y_new = y_new.loc[common_index_new]

# --- Step 3: Make predictions using the previously trained Random Forest model ---
# Ensure model_rf is available from previous execution
if 'model_rf' not in locals():
    print("Error: Random Forest model (model_rf) not found. Please run the training cell first.")
elif X_new.empty:
    print("Error: No data available for prediction after feature engineering. Check the date range or data availability.")
else:
    y_pred_new = model_rf.predict(X_new)

    # --- Step 4: Report Performance Metrics for the new period ---
    accuracy_new = accuracy_score(y_new, y_pred_new)
    conf_matrix_new = confusion_matrix(y_new, y_pred_new)
    class_report_new = classification_report(y_new, y_pred_new)

    print(f"\n### Random Forest Classifier Performance Metrics (2023-01-01 to 2024-01-01):\n")
    print(f"- **Accuracy:** {accuracy_new:.2f}")
    print(f"- **Confusion Matrix:**\n{conf_matrix_new}")
    print(f"- **Classification Report:**\n{class_report_new}")


/tmp/ipykernel_3247/301875025.py:13: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data_new = yf.download("AAPL", start=start_date_new, end=end_date_new)
[*********************100%***********************]  1 of 1 completed


### Random Forest Classifier Performance Metrics (2023-01-01 to 2024-01-01):

- **Accuracy:** 0.47
- **Confusion Matrix:**
[[80 12]
 [94 15]]
- **Classification Report:**
              precision    recall  f1-score   support

           0       0.46      0.87      0.60        92
           1       0.56      0.14      0.22       109

    accuracy                           0.47       201
   macro avg       0.51      0.50      0.41       201
weighted avg       0.51      0.47      0.39       201



In [ ]:
# Based on the performance metrics for the period from 2023-01-01 to 2024-01-01,
# the Random Forest classifier showed an Accuracy of 47.26%. This is quite low
# and suggests the model is not very effective at predicting stock price movements
# for this new period.

# Let's break down the Classification Report:
# - Precision for class 0 (price does not increase): 0.46. This means that when
#   the model predicts the price won't increase, it's correct 46% of the time.
# - Recall for class 0 (price does not increase): 0.87. This means the model
#   correctly identifies 87% of the actual instances when the price does not increase.
# - Precision for class 1 (price increases): 0.56. This means that when the model
#   predicts the price will increase, it's correct 56% of the time.
# - Recall for class 1 (price increases): 0.14. This is very low, indicating that
#   the model only correctly identifies 14% of the actual instances when the price
#   *does* increase. It misses a large majority of upward movements.

# Confusion Matrix:
# [[80, 12],
#  [94, 15]]
# - True Negatives (80): Correctly predicted price would not increase.
# - False Positives (12): Incorrectly predicted price would increase (when it actually didn't).
# - False Negatives (94): Incorrectly predicted price would not increase (when it actually did).
# - True Positives (15): Correctly predicted price would increase.

# The high number of False Negatives (94) compared to True Positives (15)
# for price increases clearly shows the model struggles to identify upward trends.

# Overall Assessment: The model's performance is poor, especially in predicting
# price increases. An accuracy of 47% is effectively worse than a coin flip
# (which would be 50%), and the low recall for positive movements makes it
# unreliable for identifying buy signals.

# Trading Strategies (with extreme caution):
# Given the current performance, building a profitable trading strategy based
# solely on this model would be very risky. However, if forced to consider
# strategies, they would primarily focus on avoiding losses rather than
# generating significant gains:

# 1.  Extreme Caution/Avoidance Strategy: Due to the low overall accuracy
#     and very poor recall for positive movements, the most prudent strategy
#     would be to avoid making any trades based on this model's predictions.
#     It is not reliable enough to inform financial decisions.

# 2.  Conservative 'No-Buy' Strategy (Highly Skeptical): If you had to use it,
#     perhaps use the model's stronger recall for 'no increase' (class 0) to avoid
#     buying. If the model predicts '0' (no increase), you might decide not to
#     enter a long position. However, even this is precarious because its precision
#     for '0' is also low (46%), meaning it will still incorrectly predict 'no increase'
#     a significant portion of the time.

# 3.  Contrarian Strategy (Very Risky): Given the model's poor performance,
#     a highly speculative and risky contrarian strategy *might* involve doing the
#     opposite of what the model predicts, particularly for 'price increase' signals.
#     However, this is more of a gamble than a strategy, as there's no guarantee that
#     the model is consistently wrong in a predictable way.

# Recommendation: The model needs significant improvement in feature engineering,
# model architecture, or hyperparameter tuning before it can be considered for any
# real-world trading strategy. Currently, it lacks the predictive power needed for
# financial decision-making.